### PR TITLE
Cleanup/show state

### PR DESCRIPTION
**Pull request by DimCitus**
That's WIP. Ideas to finish this:

  1. [x] We should probably review the monitor side API and return the result set we want from there directly, including the ORDER BY
  2. [x] I have added the “formation” column but thinking more about it I don't think it's needed.
  3. [x] The group/node IDs might be important to have still, adding to the node name
  4. [x] We should probably have a command `pg_autoctl show state --local` that does the same as `pg_autoctl do fsm state` so that end-users may inspect the current local state of their node, whether it's running or not, for debug purposes.
  5. [x] The `pg_autoctl perform failover` and other commands (enable/disable maintenance) should have node output that looks like the `show state` output.
  6. [x] The `pg_autoctl show events` command should be cleaned too, stop showing the formation (it's the same on every line by design) and use same `0/1` format for node id that other commands.
  7. [x] Consider adding the LSN to the `pg_autoctl show state` output.
  8. [x] <del>consider having different output level of details depending on the currently selected verbosity</del>

### CODE REVIEW

**Comment by DimCitus** [comment]
Within this cleanup work, we lose the ability to show the current replication settings for the nodes in a formation. I think `pg_autoctl show state` is not the place to show that information, and that we still should have a way to display it in a condensed form for easy review.

```
$ pg_autoctl get formation properties
Context   |                    Setting   Value
----------+----------------------------+------
formation |      number_sync_standbys  |    1
primary   | synchronous_standby_names  |    *
node_1    |        candidate priority  |   50
node_1    |        replication quorum  | true
node_2    |        candidate priority  |   50
node_2    |        replication quorum  | true
```

A nested JSON representation should be available too, with an array of nodes and their properties easily available.

File: src/monitor/notifications.c [hunk 1]
```diff
@@ -74,22 +74,27 @@ NotifyStateChange(AutoFailoverNode *node, char *description)
...
+
+	appendStringInfo(payload, "\"type\": \"state\", ");
+
+	appendStringInfo(payload, "\"formation\": \"%s\", ",
+					 node->formationId);
+
+	appendStringInfo(payload, "\"groupId\": %d, ", node->groupId);
+	appendStringInfo(payload, "\"nodeId\": %d, ", node->nodeId);
+	appendStringInfo(payload, "\"name\": \"%s\", ", node->nodeName);
+	appendStringInfo(payload, "\"host\": \"%s\", ", node->nodeHost);
```
Comment (JelteF) [inline src/monitor/notifications.c:88]: This will result in invalid JSON if nodeName has `"`. Doesn't postgres have some easy to use json formatting/quoting utilities, because of JSONB? If not, why did we move away from the previous format? 

I guess this edgecase shouldn't really pop up normally, since names are unlikely to have `"` characters in them. But you never know with users, somehow they always end up doing something that breaks your stuff.

File: src/bin/pg_autoctl/cli_do_misc.c [hunk 2]
```diff
@@ -227,13 +227,19 @@ keeper_cli_create_replication_user(int argc, char **argv)
...
 
 	if (!pg_setup_init(&pgSetup, &keeperOptions.pgSetup, true, true))
 	{
 		exit(EXIT_CODE_PGCTL);
 	}
 
+	if (!pg_controldata(&pgSetup, missingPgdataOk))
+	{
+		exit(EXIT_CODE_PGCTL);
+	}
```
Comment (JelteF) [inline src/bin/pg_autoctl/cli_do_misc.c:241]: Why is this needed now?

File: src/bin/pg_autoctl/cli_enable_disable.c [hunk 3]
```diff
@@ -501,6 +501,7 @@ cli_enable_maintenance(int argc, char **argv)
 			&(keeper.monitor),
 			keeper.config.formation,
 			keeper.config.groupId,
+			keeper.config.pgSetup.pgKind,
```
Comment (JelteF) [inline src/bin/pg_autoctl/cli_enable_disable.c:504]: Maybe we should just pass in the config here at this point, same for below.

File: src/bin/pg_autoctl/cli_do_monitor.c [hunk 4]
```diff
@@ -581,7 +582,8 @@ cli_do_monitor_version(int argc, char **argv)
 static void
 cli_do_monitor_parse_notification(int argc, char **argv)
```
Comment (JelteF) [inline src/bin/pg_autoctl/cli_do_monitor.c:583]: The comment of this function is incorrect now I think.

File: src/bin/pg_autoctl/nodestate_utils.h [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+#ifndef NODESTATE_H
+#define NODESTATE_H
+
+#include <stdbool.h>
+
+#include "pgsql.h"
+
+/*
+ * CurrentNodeState gather information we retrive through the monitor
```
Comment (JelteF) [inline src/bin/pg_autoctl/nodestate_utils.h:18]:
```suggestion
 * CurrentNodeState gathers information we retrieve through the monitor
```

File: src/bin/pg_autoctl/cli_common.c [hunk 6]
```diff
@@ -1016,6 +1016,7 @@ monitor_init_from_pgsetup(Monitor *monitor, PostgresSetup *pgSetup)
 			}
 
 			*monitor = keeper.monitor;
+			*pgSetup = config.pgSetup;
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/cli_common.c:1019]: I wanted to have the the pgSetup.pgKind information available in the commands where we call this function, and I think in the last version of the path we don't use that anyway.

File: src/monitor/notifications.c (hunk 1 above)
Comment (DimCitus) [reply src/monitor/notifications.c:88]: That's a good point. I looked at mimicking more of `json_build_object` as in https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/json.c#L991  ; and then I saw other code in Postgres where they just compose the JSON entries like I did here. We could also copy paste the following code in our extension and compose the JSON notification from those pieces?

``` c
/*
 * Append JSON text for "val" to "result".
 *
 * This is just a thin wrapper around datum_to_json.  If the same type will be
 * printed many times, avoid using this; better to do the json_categorize_type
 * lookups only once.
 */
static void
add_json(Datum val, bool is_null, StringInfo result,
		 Oid val_type, bool key_scalar)
{
	JsonTypeCategory tcategory;
	Oid			outfuncoid;

	if (val_type == InvalidOid)
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
				 errmsg("could not determine input data type")));

	if (is_null)
	{
		tcategory = JSONTYPE_NULL;
		outfuncoid = InvalidOid;
	}
	else
		json_categorize_type(val_type,
							 &tcategory, &outfuncoid);

	datum_to_json(val, is_null, result, tcategory, outfuncoid, key_scalar);
}
```

File: src/bin/pg_autoctl/cli_do_misc.c (hunk 2 above)
Comment (DimCitus) [reply src/bin/pg_autoctl/cli_do_misc.c:241]: Because we don't call `pg_controldata` from `pg_setup_init` as we used to in another PR. Baring that the output of the `pg_autoctl do pgsetup discover` commands are always having zeroes for the pg_control fields.

File: src/bin/pg_autoctl/cli_enable_disable.c (hunk 3 above)
Comment (DimCitus) [reply src/bin/pg_autoctl/cli_enable_disable.c:504]: I think it's best to avoid having `monitor.c` depend on `keeper_config.c`.

File: src/monitor/notifications.c (hunk 1 above)
Comment (DimCitus) [reply src/monitor/notifications.c]: Turns out that Postgres exposes this escaping facility that we can use directly, now using it:

``` c
	escape_json(payload, node->formationId);
```

File: src/bin/pg_autoctl/cli_do_monitor.c (hunk 4 above)
Comment (DimCitus) [reply src/bin/pg_autoctl/cli_do_monitor.c:587]: Fixed.

File: src/bin/pg_autoctl/nodestate_utils.h (hunk 5 above)
Comment (DimCitus) [reply src/bin/pg_autoctl/nodestate_utils.h]: Fixed.

File: src/bin/pg_autoctl/nodestate_utils.c [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+
+	fformat(stdout, "%*s | %*s | %*s | %17s | %17s | %17s | %6s\n",
+			headers->maxNameSize, nodeState->node.name,
+			headers->maxNodeSize, composedId,
+			headers->maxHostSize, hostport,
+			NodeStateToString(nodeState->reportedState),
+			NodeStateToString(nodeState->goalState),
+			nodeState->node.lsn,
+			nodeState->health == -1 ? "?" :
+			nodeState->health == 1 ? "✓" : "✗");
```
[the remark this answers is not in the document]
Comment (JelteF) [reply src/bin/pg_autoctl/nodestate_utils.c:223]: Maybe instead of `Health` this should be called `Reachable` or something similar. It wasn't obvious to me that this `Health` didn't include anything about the current state, but simply if the monitor could reach the node. I'm not sure if it still makes sense then to put it in front of `Current State`.

File: src/bin/pg_autoctl/cli_show.c [hunk 8]
```diff
@@ -400,12 +401,121 @@ cli_show_state(int argc, char **argv)
...
+
+			nodeState.reportedState = keeper.state.current_role;
+			nodeState.goalState = keeper.state.assigned_role;
+
+			strlcpy(nodeState.node.lsn,
+					config.pgSetup.control.recoveryEndingLocation,
+					PG_LSN_MAXLENGTH);
+
+			/* we have no idea, only the monitor knows, so display "?" */
+			nodeState.health = -1;
```
[the remark this answers is not in the document]
Comment (JelteF) [reply src/bin/pg_autoctl/cli_show.c:470]: I misunderstood the meaning of `Health` initially. We know the node is unhealthy if postgres is down then.

File: src/bin/pg_autoctl/cli_show.c [hunk 9]
```diff
@@ -400,12 +401,121 @@ cli_show_state(int argc, char **argv)
...
+			}
+
+			if (!keeper_init(&keeper, &config))
+			{
+				/* errors have already been logged */
+				exit(EXIT_CODE_BAD_CONFIG);
+			}
+
+			/* also grab the minimum recovery LSN if that's possible */
+			if (!pg_controldata(&(config.pgSetup), missingPgdataIsOk))
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/cli_show.c:448]: I'm not sure which of the available LSN locations we should display here. That said you're right, the current one that I've chosen is confusing for the user. I'm going to switch to using the "Last checkpoint location" which seems less confusing from a quick local test:

```
$ pg_autoctl show state
Name |  Node |      Host:Port |     Current State |    Assigned State |               LSN | Health
-----+-------+----------------+-------------------+-------------------+-------------------+-------
   a |     1 | localhost:5001 |           primary |           primary |         0/5000640 |    ✓
   b |     2 | localhost:5002 |         secondary |         secondary |         0/5000640 |    ✓
   c |     3 | localhost:5003 |         secondary |         secondary |         0/5000640 |    ✓

$ pg_controldata /tmp/pgaf/c | grep location
Latest checkpoint location:           0/4000060
Latest checkpoint's REDO location:    0/4000028
Minimum recovery ending location:     0/4000100
Backup start location:                0/0
Backup end location:                  0/0
```

File: src/bin/pg_autoctl/cli_show.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/cli_show.c:470]: Well, even when we're DEMOTED, we can't know what's the current health value on the monitor if we can't contact it. It might be stuck in the previous connection attempt at the moment when we do `pg_autoctl show state --local`, so I assume that we just don't know.

File: src/bin/pg_autoctl/nodestate_utils.c [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+	json_object_set_string(jsobj, "current_group_state",
+						   NodeStateToString(nodeState->reportedState));
+
+	json_object_set_string(jsobj, "assigned_group_state",
+						   NodeStateToString(nodeState->goalState));
+
+	json_object_set_string(jsobj, "Minimum Recovery Ending LSN",
+						   nodeState->node.lsn);
+
+	json_object_set_boolean(jsobj, "health", nodeState->health == 1);
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/nodestate_utils.c:303]: Or maybe we could just expose the value on the monitor as it is, an integer, -1, 0, or 1? I wanted to simplify it down to a yes/no, but if you think that's not helpful to the user, I'm not sure we should “translate” our internal representation there...

File: src/monitor/notifications.c [hunk 11]
```diff
@@ -74,22 +75,32 @@ NotifyStateChange(AutoFailoverNode *node, char *description)
...
+	appendStringInfo(payload, ", \"host\": ");
+	escape_json(payload, node->nodeHost);
+
+	appendStringInfo(payload, ", \"port\": %d, ", node->nodePort);
+
+	appendStringInfo(payload, "\"reportedState\": ");
+	escape_json(payload, ReplicationStateGetName(node->reportedState));
+
+	appendStringInfo(payload, ", \"goalState\": ");
+	escape_json(payload, ReplicationStateGetName(node->goalState));
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/monitor/notifications.c:101]: Good point. I don't think we would use it for now, but it's easy to do thanks to using JSON.

File: src/bin/pg_autoctl/nodestate_utils.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (JelteF) [reply src/bin/pg_autoctl/nodestate_utils.c:303]: Also fine by me

File: src/bin/pg_autoctl/parsing.c [hunk 12]
```diff
@@ -216,119 +224,123 @@ parse_controldata_field_uint64(const char *controlDataString,
...
 	{
-		log_warn("Failed to parse node id \"%s\"", ptr);
+		log_error("Failed to parse formation in JSON "
+				  "notification message \"%s\"", message);
 		return false;
 	}
-	ptr = ++col;
+	strlcpy(nodeState->formation, str, sizeof(nodeState->formation));
+
+	number = json_object_get_number(jsobj, "groupId");
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/parsing.c:294]: The lib parson API isn't well tented for error handling. When you want to handle such errors you need to go down an abstraction level from the object to fetching its value, and then working with the value API, which contains:

``` c
double json_value_get_number(const JSON_Value *value) {
    return json_value_get_type(value) == JSONNumber ? value->value.number : 0;
}
```

As elsewhere in the code, I have continued to implement a model wherein we trust the output we get from the monitor.

File: src/bin/pg_autoctl/nodestate_utils.h [hunk 13]
```diff
@@ -15,7 +15,7 @@
 #include "pgsql.h"
 
 /*
- * CurrentNodeState gather information we retrive through the monitor
+ * CurrentNodeState gathers information we retrive through the monitor
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/nodestate_utils.h]: Fixed, thanks.

File: src/monitor/notifications.c [hunk 14]
```diff
@@ -79,20 +80,25 @@ NotifyStateChange(AutoFailoverNode *node, char *description)
 
 	appendStringInfo(payload, "\"type\": \"state\", ");
 
-	appendStringInfo(payload, "\"formation\": \"%s\", ",
-					 node->formationId);
+	appendStringInfo(payload, "\"formation\": ");
+	escape_json(payload, node->formationId);
 
-	appendStringInfo(payload, "\"groupId\": %d, ", node->groupId);
+	appendStringInfo(payload, ", \"groupId\": %d, ", node->groupId);
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/monitor/notifications.c]: Good point, done now.

File: src/monitor/notifications.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/monitor/notifications.c:101]: Added the health as a string that maps to the ENUM, so the values are "unknown", "bad", or "good". Not used client-side at the moment.

File: src/bin/pg_autoctl/nodestate_utils.c [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+	char hostport[BUFSIZE] = { 0 };
+	char composedId[BUFSIZE] = { 0 };
+
+	(void) nodestatePrepareNode(headers,
+								&(nodeState->node),
+								nodeState->groupId,
+								hostport,
+								composedId);
+
+	fformat(stdout, "%*s | %*s | %*s | %17s | %17s | %17s | %6s\n",
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/nodestate_utils.c]: The state is now a `%18s` when it used to be a `%17s`, and when refactoring the code I mistakenly kept the `%17s` that applies to the LSN width. Now fixed, and LSN width is computed dynamically too.

I don't think we should compute the state widths dynamically: one of the use cases for `pg_autoctl show state` is when used with a `watch -n 1` command, and the states are going to change. It would be distracting if each output from the watch command was of a different line length. The other fields (name, host:port, lsn) that are computed dynamically will change less often, and I think it's okay to by all dynamic there.

File: src/bin/pg_autoctl/nodestate_utils.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/nodestate_utils.c]: In terms of user experience it's not the best, because of unicode character display width. Even with a monospace font, chars such as ✓ and ✗ will take a different width when printed on the terminal. This will then break our attempt at having any alignment at all in the output for most users. That's the reason why this is the only column with unicode, and it has to be the last field on the output.

File: src/bin/pg_autoctl/nodestate_utils.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/nodestate_utils.c]: Done that way now.

File: src/bin/pg_autoctl/nodestate_utils.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (JelteF) [reply src/bin/pg_autoctl/nodestate_utils.c]: It sohuld be `%19s`, there's two characters too little per column in the output I have above. Rest makes sense.

File: src/bin/pg_autoctl/cli_show.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (JelteF) [reply src/bin/pg_autoctl/cli_show.c:448]: maybe if postgres is up we can ask it using one of the UDFs. If it's down, I guess it would be one of the latest checkpoint location reltaed ones.

File: src/bin/pg_autoctl/parsing.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (JelteF) [reply src/bin/pg_autoctl/parsing.c:296]: Makes sense, as long as it doesn't start reading uninitialized memory I think it's indeed fine here. And like you show it simply returns 0 in other cases.

File: src/bin/pg_autoctl/nodestate_utils.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (JelteF) [reply src/bin/pg_autoctl/nodestate_utils.c]: What do you think of renaming he column to `Reachable` instead of `Healthy` for clarity?

Also, we could simply write them out in words as well. `Yes`, `No` and `?` instead of using unicode. If that's causing issues.

File: src/bin/pg_autoctl/cli_show.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/cli_show.c:448]: My understanding is that should we restart the secondary we will see the LSN get in the past and forward again after crash recovery is done.  We might want to change the column name when using `--local`, also.

File: src/bin/pg_autoctl/nodestate_utils.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/nodestate_utils.c]: Ah yeah. I have now made it internally "dynamic" even though forced to MAX_NODE_STATE_LEN always, and this constant is now edited in the `state.h` header, which is where we add new states. Should make it easier to maintain in the future.

File: src/bin/pg_autoctl/nodestate_utils.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/nodestate_utils.c]: Renamed to Reachable, that's a good point, and then used the strings "unknown", "yes", and "no". Makes it easier.

File: src/bin/pg_autoctl/nodestate_utils.c [hunk 16]
```diff
@@ -139,17 +145,24 @@ nodestateAdjustHeaders(NodeAddressHeaders *headers,
...
+	 * always use the max known state length.
+	 */
+	headers->maxStateSize = MAX_NODE_STATE_LEN;
+
 	/* initialize to mininum values, if needed */
 	if (headers->maxNameSize == 0)
 	{
 		/* Name */
-		headers->maxNameSize = 5;
+		headers->maxNameSize = strlen("Name");
```
Comment (JelteF) [inline src/bin/pg_autoctl/nodestate_utils.c:159]: I like this self documenting code

File: src/bin/pg_autoctl/nodestate_utils.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/nodestate_utils.c]: And now moved the Reachable field before the State field, with the headers being all "computed" it's easy now.

File: src/bin/pg_autoctl/nodestate_utils.c (hunk 16 above)
Comment (DimCitus) [reply src/bin/pg_autoctl/nodestate_utils.c:159]: I just grew fed-up of having to review and maintain those numbers everywhere...

File: src/bin/pg_autoctl/nodestate_utils.c [hunk 17]
```diff
@@ -0,0 +1,374 @@
...
+	json_object_set_string(jsobj, "current_group_state",
+						   NodeStateToString(nodeState->reportedState));
+
+	json_object_set_string(jsobj, "assigned_group_state",
+						   NodeStateToString(nodeState->goalState));
+
+	json_object_set_string(jsobj, "Minimum Recovery Ending LSN",
+						   nodeState->node.lsn);
+
+	json_object_set_number(jsobj, "health", (double) nodeState->health);
```
Comment (JelteF) [inline src/bin/pg_autoctl/nodestate_utils.c:342]: Maybe nice to use `nodestateHealthToString` here as well.

Comment (DimCitus) [reply src/bin/pg_autoctl/nodestate_utils.c]: Done now.

File: src/bin/pg_autoctl/cli_show.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/cli_show.c]: We're now using the LSN obtained connecting to Postgres when it's running, so that's the same LSN that we send to the monitor. When Postgres is not running we grab the Latest checkpoint location from the output of pg_controldata.
